File: 121_best_time_buy_and_sell/maxProfit.js
var maxProfit = function (prices) {
  if (prices === null || prices.length === 0) {
    return 0;
  }
  let profit = 0;
  let minPrice = prices[0];
  prices.forEach((element) => {
    profit = Math.max(profit, element - minPrice);
    minPrice = Math.min(minPrice, element);
  });
  return profit;
};

File: 121_best_time_buy_and_sell/MaxProfit.java
public class MaxProfit {
    public static int maxProfit(int[] prices) {
        if (prices == null || prices.length == 0) {
            return 0;
        }
        int profit = 0;
        int min = prices[0];
        for (int i = 0; i < prices.length; i++) {
            profit = Math.max(profit, prices[i] - min);
            min = Math.min(min, prices[i]);
        }

        return profit;
    }

    public static void main(String[] args) {
        int[] prices = { 7, 1, 5, 3, 6, 4 };
        // int[] prices = { 7, 6, 4, 3, 1 };
        int result = maxProfit(prices);
        System.out.println(result);
    }
}
